src/v/cloud_storage_clients/configuration.cc

```diff
@@ -170,4 +171,45 @@ std::ostream& operator<<(std::ostream& o, const abs_configuration& c) {
     return o;
 }
 
+model::cloud_storage_backend infer_backend_from_configuration(
+  const client_configuration& client_config,
+  model::cloud_credentials_source cloud_storage_credentials_source) {
+    if (auto v = config::shard_local_cfg().cloud_storage_backend.value();
+        v != model::cloud_storage_backend::unknown) {
+        return v;
+    }
+
+    if (std::holds_alternative<abs_configuration>(client_config)) {
+        return model::cloud_storage_backend::azure;
+    }
+
+    if (
+      cloud_storage_credentials_source
+      != model::cloud_credentials_source::config_file) {
+        switch (cloud_storage_credentials_source) {
+        case model::cloud_credentials_source::aws_instance_metadata:
+            [[fallthrough]];
+        case model::cloud_credentials_source::sts:
+            return model::cloud_storage_backend::aws;
+        case model::cloud_credentials_source::gcp_instance_metadata:
+            return model::cloud_storage_backend::google;
+        case model::cloud_credentials_source::config_file:
+            __builtin_unreachable();
```

Comment:
@abhijat 

please use an assertion for cases like this where you are expressing that logically you should never reach this point. if you use builtin_unreachable here and we end up crashing we'll get a bad instruction segmentation fault which wont' provide much helpful context on its own.

builtin_unreachable in these cases is just for convincing gcc that it shouldn't warn about switch statements not covering all the possible conditions when they are actually covered.

Reply:
sure, added in the follow up PR https://github.com/redpanda-data/redpanda/pull/9252 latest commit https://github.com/redpanda-data/redpanda/pull/9252/commits/432e54d5d00006408c328179928c567a25b42782